### DISCUSSION TITLE
Optimizing .take() for sample-based access on large, sparse time-series data?

### DISCUSSION THREAD

**Discussion by lyang2012** [lance-format/lance#5704] [Q&A]
I’m working with a large, sparse, multi-dimensional time-series dataset stored in Lance, and I’d like feedback on whether I’m using Lance in the intended way and how to further optimize performance.

**Dataset layout**

Each row looks like:
`id (string), date_idx (int32), region_idx (int32), feature_1 (float32), feature_2 (float32)`

A logical sample (identified by id) spans multiple rows across (date_idx, region_idx).
Different samples have different numbers of rows because the data is sparse along both dimensions.

To optimize access:

The dataset is sorted by id, so all rows for a sample are stored contiguously.

I also maintain a sample index metadata table (stored separately as Parquet) that maps each sample to:
- start_row
- row_length

This allows me to convert sample indices → row indices efficiently.

**Access pattern:**

At training time, I randomly sample batches of samples. For each batch:

1. Convert sample indices to row indices
2. Sort row indices to preserve contiguity
3. Call .take() to fetch the rows
4. “Tensify” the sparse rows into a dense tensor of shape (num_samples, num_dates, num_regions, num_features)

Relevant code snippets below:

```
def get_samples_row_indices(sample_map, sample_indices):
    all_row_indices = []
    rows_per_sample = []
    sorted_indices = sorted(sample_indices)
    for idx in sorted_indices:
        start, length = sample_map[idx]
        all_row_indices.extend(range(start, start + length))
        rows_per_sample.append(length)
    return all_row_indices, rows_per_sample


def load_sample_map_parquet(path):
    table = pq.read_table(path)
    return list(zip(table["start"].to_pylist(), table["length"].to_pylist()))


def tensify(table: pa.Table, num_samples: int, rows_per_sample: list[int], num_dates: int, num_regions: int, feature_cols: list[str]):
    rows_per_sample = np.array(rows_per_sample)
    sample_idx = np.repeat(np.arange(len(rows_per_sample)), rows_per_sample)

    date_idx = table["date_idx"].to_numpy()
    region_idx = table["region_idx"].to_numpy()

    features = np.stack([table[c].to_numpy(zero_copy_only=False) for c in feature_cols], axis=-1).astype(np.float32)
    num_features = len(feature_cols)

    tensor = torch.zeros(num_samples, num_dates, num_regions, num_features, dtype=torch.float32)
    indices = (
        torch.from_numpy(sample_idx).long(),
        torch.from_numpy(date_idx).long(),
        torch.from_numpy(region_idx).long()
    )
    tensor[indices] = torch.from_numpy(features)

    return tensor


ds = lance.dataset("data_path/data")
len(ds) # 346_358_730 rows

sample_index_map = load_sample_map_parquet("/data_path/metadata/sample_index.parquet")
len(sample_index_map) # 379_170 samples

sampler = RandomSampler(range(len(sample_index_map)))
batch_sampler = BatchSampler(sampler, batch_size=512, drop_last=False)


num_batches = 0
total_indexing_time = 0 
total_decoding_time = 0
total_tensifying_time = 0
total_value = 0
for indices in batch_sampler:
    start_indexing = time.time()
    row_indcies, rows_per_sample = get_samples_row_indices(sample_index_map, indices)
    total_indexing_time += (time.time() - start_indexing)
    
    start_decoding = time.time()
    sub_table = ds1.take(row_indcies, columns=["region_idx", "date_idx", "feature1"])
    total_decoding_time += (time.time() - start_decoding)

    start_tensifying = time.time()
    sorted_tensor = tensify(sub_table, num_samples=len(indices), rows_per_sample=rows_per_sample, num_dates=64, num_regions=183, feature_cols=["feature1"])
    unsorted_tensor = sorted_tensor[np.argsort(np.argsort(indices))]
    total_tensifying_time += (time.time() - start_tensifying)

    
    num_batches += 1

print(f"Num batches: {num_batches}")
print(f"Indexing time: {total_indexing_time} | Time per batch: {total_indexing_time/num_batches:.4f}")
print(f"Decoding time: {total_decoding_time} | Time per batch: {total_decoding_time/num_batches:.4f}")
print(f"Tensifying time: {total_tensifying_time} | Time per batch: {total_tensifying_time/num_batches:.4f}")
```
**Scale**
- Dataset size: 346,358,730 rows
- Number of samples: 379,170
- Batch size: 512 samples
- Average rows per batch: ~467K

**Benchmark results**
```
Num batches: 741
Indexing time: 1.8271229267120361 | Time per batch: 0.0025
Decoding time: 34.384377241134644 | Time per batch: 0.0464
Tensifying time: 4.653211832046509 | Time per batch: 0.0063
```

Fetching ~467K rows in ~46 ms is not terrible, but `.take()` is clearly the dominant cost.

I’ve also observed that:

- When the dataset is cached (no disk IO), the .take() runtime is roughly unchanged.
- This suggests the bottleneck is CPU-bound, likely due to decoding / reassembly rather than IO?

**Questions**

1. Does this data layout and access pattern align with how Lance is intended to be used for sample-based workloads?
2. Are there recommended ways to further optimize .take() for this pattern (e.g., layout choices, fragment sizing, alternative APIs)?
3. Is there a more “Lance-native” way to represent or access sparse multi-dimensional samples like this?

Any feedback or suggestions would be greatly appreciated. Thanks!

**Comment by westonpace** [comment]
> Does this data layout and access pattern align with how Lance is intended to be used for sample-based workloads?

Yes.

> This suggests the bottleneck is CPU-bound, likely due to decoding / reassembly rather than IO?

That's one possibility.  Though decoding and reassembly are generally pretty quick.  One way to get a good start on understanding the bottlenecks is to turn on tracing by doing two things.  First, put this at the top of your script.

```
from lance.tracing import trace_to_chrome
trace_to_chrome(file="/path/to/trace.json")
```

Second, set `LANCE_LOG=debug` (in newer pre-releases you can set `LANCE_TRACING=debug` instead and don't have to have all the logging output you'd normally get)

Then you can look at the trace files that get generated at https://ui.perfetto.dev .  Or, if you post a trace file here, we can look at it for you.

> Are there recommended ways to further optimize .take() for this pattern (e.g., layout choices, fragment sizing, alternative APIs)?

Sorting the indices in advance is good.  Down at the rust level the file reader APIs actually take in a list of ranges.  It's possible that exposing those APIs up to python may help your performance (e.g. so you can pass one range instead of ~900 indices).  Though this might be moot (see next point).

> Is there a more “Lance-native” way to represent or access sparse multi-dimensional samples like this?

What makes these sparse?

As for multi-dimensional, you can store the samples in a `List<Struct>`.  For example, something like...

```
>>> pa.table({"sample_id": [0, 1], "samples": [{"date_idx": 1, "region_idx": 1}, {"date_idx": 1, "region_idx": 2}]})
pyarrow.Table
sample_id: int64
samples: struct<date_idx: int64, region_idx: int64>
  child 0, date_idx: int64
  child 1, region_idx: int64
----
sample_id: [[0,1]]
samples: [
  -- is_valid: all not null
  -- child 0 type: int64
[1,1]
  -- child 1 type: int64
[1,2]]
```

The main advantage here is going to be the same as above.  Instead of providing 900 row indices per sample you'd only need to provide one.

Regardless of which approach you try you might also try replacing the mapping table with a btree index on the sample id.  Then you can do:

```
indices_str = indices.join(",")
ds.to_table(columns=["region_idx", "date_idx", "feature1"], filter=f"id in ({indices_str})")
```

Not sure if this will help performance at all.

**Comment by omni-front** [comment]
you might want to look into optimizing the `.take()` operation itself. sometimes the bottleneck is in the conversion between row indices and actual data fetching. also, double-check the tensification step to ensure it's not introducing unnecessary overhead. if you're already doing these, not sure what else without diving deeper into the specific setup.
